source/common/ext_authz/ext_authz_impl.cc

```diff
@@ -0,0 +1,250 @@
+
+GrpcClientImpl::~GrpcClientImpl() { ASSERT(!callbacks_); }
+
+void GrpcClientImpl::cancel() {
+  ASSERT(callbacks_ != nullptr);
+  request_->cancel();
+  callbacks_ = nullptr;
+}
+
+void GrpcClientImpl::check(RequestCallbacks& callbacks, const envoy::api::v2::auth::CheckRequest& request,
+                           Tracing::Span& parent_span) {
+  ASSERT(callbacks_ == nullptr);
+  callbacks_ = &callbacks;
+
+  request_ = async_client_->send(service_method_, request, *this, parent_span, timeout_);
+}
+
+void GrpcClientImpl::onSuccess(std::unique_ptr<envoy::api::v2::auth::CheckResponse>&& response,
+                               Tracing::Span& span) {
+  CheckStatus status = CheckStatus::OK;
+  ASSERT(response->status().code() != Grpc::Status::GrpcStatus::Unknown);
+  if (response->status().code() != Grpc::Status::GrpcStatus::Ok) {
+    status = CheckStatus::Denied;
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceUnauthz);
+  } else {
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceOk);
+  }
+
+  callbacks_->complete(status);
+  callbacks_ = nullptr;
+}
+
+void GrpcClientImpl::onFailure(Grpc::Status::GrpcStatus status, const std::string&,
+                               Tracing::Span&) {
+  ASSERT(status != Grpc::Status::GrpcStatus::Ok);
+  UNREFERENCED_PARAMETER(status);
+  callbacks_->complete(CheckStatus::Error);
+  callbacks_ = nullptr;
+}
+
+GrpcFactoryImpl::GrpcFactoryImpl(const std::string& cluster_name,
+                                 Upstream::ClusterManager& cm)
+    : cluster_name_(cluster_name), cm_(cm) {
+  if (!cm_.get(cluster_name_)) {
+    throw EnvoyException(fmt::format("unknown external authorization service cluster '{}'", cluster_name_));
+  }
+}
+
+ClientPtr GrpcFactoryImpl::create(const Optional<std::chrono::milliseconds>& timeout) {
+  return ClientPtr{new GrpcClientImpl(
+      ExtAuthzAsyncClientPtr{
+          new Grpc::AsyncClientImpl<envoy::api::v2::auth::CheckRequest,
+                                    envoy::api::v2::auth::CheckResponse>(cm_, cluster_name_)},
+      timeout)};
+}
+
+::envoy::api::v2::Address* CheckRequestGen::get_pbuf_address(const Network::Address::InstanceConstSharedPtr& instance) {
+  using ::envoy::api::v2::Address;
+
+  Address *addr = new Address();
+  ASSERT(addr);
+
+  if (instance->type() == Network::Address::Type::Ip) {
+    addr->mutable_socket_address()->set_address(instance->ip()->addressAsString());
+    addr->mutable_socket_address()->set_port_value(instance->ip()->port());
+  } else {
+    ASSERT(instance->type() == Network::Address::Type::Pipe);
+    addr->mutable_pipe()->set_path(instance->asString());
+  }
+  return addr;
+}
+
+::envoy::api::v2::auth::AttributeContext_Peer* CheckRequestGen::get_connection_peer(const Network::Connection *connection, const std::string& service, const bool local) {
+  using ::envoy::api::v2::auth::AttributeContext_Peer;
+
+  AttributeContext_Peer *peer = new AttributeContext_Peer();
```

Comment:
Please use `unique_ptr` rather than managing naked pointers when doing ownership transfer.

Reply:
Will do. Learning about smart pointers with this PR!!